### ISSUE TITLE
Problem with cameras. Offset is None

### ISSUE THREAD

**Issue by Andreus00** [eth-ait/4d-dress#3]
Hi,

First of all, thank you for your work!!!

I am trying to load camera parameters from the _cameras.pkl_ file, and the mesh as a point cloud to use with Gaussian Splatting.

To do so, I wrote this:
`

def read4dDressInfo(cfg) -> SceneInfo:

    # open info from basic_info.pkl
    basic_info = open_pickle(cfg.source_path, "basic_info.pkl")
    scan_frames = basic_info['scan_frames']
    scan_rot = basic_info['rotation']

    offset = basic_info['offset']

    # load scan mesh
    _, scan_mesh, _, _ = load_scan_mesh(cfg.obj_path, rotation=scan_rot, offset=offset)

    # load cameras from cameras.pkl
    cameras = open_pickle(cfg.source_path, "Capture", "cameras.pkl")
    
    # For each camera, read image, mask and camera.
    cam_infos = []

    # For each camera, create a CameraInfo object
    for key, camera in cameras.items():
         
        camera_intrinsics = camera['intrinsics']
        camera_extrinsics = camera['extrinsics']

        R = camera_extrinsics[:3, :3]
        R = np.transpose(R)
        T = camera_extrinsics[:3, 3]
        
        image_path = os.path.join(cfg.source_path, "Capture", key, "images", f"capture-f00011.png")
        image = open_image(image_path)
        mask_path = os.path.join(cfg.source_path, "Capture", key, "masks", f"mask-f{scan_frames[0]}.png")
        mask = open_image(mask_path)
 
        f_x = camera_intrinsics[0, 0]
        f_y = camera_intrinsics[1, 1]
        width = image.size[0]
        height = image.size[1]

        fov_x = 2 * np.arctan(width / (2 * f_x))
        fov_y = 2 * np.arctan(height / (2 * f_y))

        alpha_image = Image.new("RGBA", image.size, (255, 255, 255, 255))
        alpha_image.paste(image, (0, 0), mask)
        print(np.asarray(alpha_image).shape)

        cam_info = CameraInfo(
            uid=key,
            R=R,
            T=T,
            FovY=fov_y,
            FovX=fov_x,
            image=alpha_image,
            image_path=image_path,
            image_name=f"side-{key}-capture-f{scan_frames[0]}",
            width=width,
            height=height
        )
        cam_infos.append(cam_info)

    # get nerf normalization
    nerf_normalization = getNerfppNorm(cam_infos)

    # create a point cloud from the mesh. This simply takes vertices and normals from the mesh and packs them into a BasicPointCloud object. The colors of Gaussians are set to normals.
    scales, opacity, pcd = mesh_to_pointcloud(scan_mesh)

    images_dir = os.path.join(cfg.source_path,cfg.subj,cfg.outfit,cfg.seq)
    ply_path = os.path.join(images_dir, "points3d.ply")

    return scales, opacity, SceneInfo(
        point_cloud=pcd,
        train_cameras=cam_infos,
        test_cameras=[],
        nerf_normalization=nerf_normalization,
        ply_path=ply_path
    )

`

However, I am having a problem as the point cloud (created from the mesh) and the image are misaligned. I plotted both the image and the render from GS, and these are some examples of the misalignment:
![immagine](https://github.com/eth-ait/4d-dress/assets/62235503/83ad6c6d-dbb6-442b-9bbf-f0e5609f5754)
![immagine](https://github.com/eth-ait/4d-dress/assets/62235503/1bdc389b-f871-459c-b1bf-ce986b993666)

If I remove the rotation from the mesh, both the subjects look in the same direction, but they remain misaligned:
![immagine](https://github.com/eth-ait/4d-dress/assets/62235503/03ecfc59-5489-4187-8cdd-07f8503220fe)
![immagine](https://github.com/eth-ait/4d-dress/assets/62235503/6d046711-3ede-425a-ab0b-19ca07430070)

Am I missing something?

Thank you.

p.s. I noticed that basic_info['offset'] is always None for the sample that I am using (0112 - Inner - Take2). Maybe that's the problem?

**Comment by Andreus00** [comment]
Update: By setting mcentral to False, the alignment improves a lot, but it still seems like all the cameras are a little bit to the left:
![immagine](https://github.com/eth-ait/4d-dress/assets/62235503/cb355332-cc36-41e8-81ef-d63501db98e2)
![immagine](https://github.com/eth-ait/4d-dress/assets/62235503/52260d75-f776-49a4-9a2f-7a0f3918aeea)

**Comment by azuxmioy** [comment]
Hi all, thanks for the message.

 I might need some time to investigate this issue due to the upcoming CVPR conference. 

At the same time, if @WenbWa has any ideas, please feel free to comment.

Thanks.

**Comment by azuxmioy** [comment]
btw, what rendering pipeline are you using?

have you tried our demo code here?
https://github.com/eth-ait/4d-dress/blob/baf3e8f0857f7b22996512ba82a55c9530f268ce/dataset/extract_garment.py

**Comment by WenbWa** [comment]
Hi, thanks for being interested in our 4D-DRESS dataset.

The cameras.pkl files are used in our work mostly for the Pytorch3D cameras.
Here is the Pytorch3D camera loading function used in dataset/extract_garment.py:
```
# load pytorch3d cameras from parameters: intrinsics, extrinsics
def load_pytorch_cameras(camera_params, camera_list, image_shape):
    # init camera_dict
    camera_dict = dict()
    # process all camera within camera_list
    for camera_id in camera_list:
        # assign camera intrinsic and extrinsic matrices
        intrinsic = torch.tensor((camera_params[camera_id]["intrinsics"]), dtype=torch.float32).cuda()
        extrinsic = torch.tensor(camera_params[camera_id]["extrinsics"], dtype=torch.float32).cuda()
        # assign camera image size
        image_size = torch.tensor([image_shape[0], image_shape[1]], dtype=torch.float32).unsqueeze(0).cuda()

        # assign camera parameters
        f_xy = torch.cat([intrinsic[0:1, 0], intrinsic[1:2, 1]], dim=0).unsqueeze(0)
        p_xy = intrinsic[:2, 2].unsqueeze(0)
        R = extrinsic[:, :3].unsqueeze(0)
        T = extrinsic[:, 3].unsqueeze(0)
        # coordinate system adaption to PyTorch3D
        R[:, :2, :] *= -1.0
        # camera position in world space -> world position in camera space
        T[:, :2] *= -1.0
        R = torch.transpose(R, 1, 2)  # row-major
        # assign Pytorch3d PerspectiveCameras
        camera_dict[camera_id] = PerspectiveCameras(focal_length=f_xy, principal_point=p_xy, R=R, T=T, in_ndc=False, image_size=image_size).cuda()
    # assign Pytorch3d RasterizationSettings
    raster_settings = RasterizationSettings(image_size=image_shape, blur_radius=0.0, faces_per_pixel=1, max_faces_per_bin=80000)
    return camera_dict, raster_settings

```

Maybe the camera parameters used in Gaussian Splatting are slightly different from the Pytorch3D ones.

**Comment by zjchu** [comment]
@Andreus00 Hi, I'm experiencing the same issue where images rendered with the existing camera parameters (from cameras.pkl ) are not aligned with the ground truth. Have you found a solution to this problem?

**Comment by yuyujunjun** [comment]
Thanks! I’m glad the explanation is helpful. The different camera behaviors across libraries and representations are always quite annoying, and it’s hard to have a single conversion method that’s reusable. That’s why I tend to organize everything in matrix form and use the most straightforward approach for coordinate transformations. I hope my code can help anyone who, like me, prefers not to rely on PyTorch3D rendering.

Besides, you’ve provided an excellent dataset and corresponding code, so I don’t think any updates are needed.

**Comment by yuyujunjun** [comment]
Correction / Clarification

Sorry, I realized I made a mistake in the previous code. The z-axis flipping should happen in the projection matrix, rather than in the model-view (MV) matrix.

Specifically: the previous code multiplies `C_p3d2nvd` with world_view_transform forapplying the z-flip:

```full_transform=proj⋅(flip_z⋅mv)```

Although this does not change the full transformation mathematically, it affects Gaussian Splatting rendering, because the rasterizer relies on the MV matrix for frustum clipping.

The correct approach is to apply the z-flip in the projection matrix:

```full_transform=(proj⋅flip_z)⋅mv```

So `C_p3d2nvd` should be multiplied by `intrinsic_to_projection()` , not by the modified MV matrix.
